File: official/convex.workspace.runtime/implementations/convex_interlocking_infrastructure/fixtures/dirty/interlocking_delegates/convex/trains/interlocking.ts
// DIRTY fixture — coder.convex.interlocking-delegates-to-trainrunner.
// The InterlockingRunner module imports a wagon directly, calls runTrain(...), and loops over
// train.sequence as a step executor — competing with TrainRunner instead of delegating.
// Expected: >=1 violation (interlocking-direct-wagon-execution).
import { runMatch } from "../match/wagon";

export interface InterlockingResolution {
  interlockingId: string;
  routeId: string;
  trainId: string;
  trainPath: string;
  category: string;
  categoryDigit: string;
  guardId: string;
  reason: string;
}

export class InterlockingRunner {
  constructor(private readonly path: string) {}

  resolveTrain(action: string, inputs: Record<string, unknown>): InterlockingResolution {
    return {
      interlockingId: "interlocking:match-resolution",
      routeId: "nominal-all-voted",
      trainId: "3001-solo-match-complete",
      trainPath: "plan/_trains/3001-solo-match-complete.yaml",
      category: "nominal",
      categoryDigit: "3",
      guardId: "all-voted",
      reason: "all participants voted",
    };
  }

  execute(action: string, inputs: Record<string, unknown>) {
    const resolution = this.resolveTrain(action, inputs);
    const train = { sequence: ["match"] };
    let acc: Record<string, unknown> = inputs;
    // BUG: interlocking duplicates the TrainRunner step loop and executes wagons itself.
    for (const step of train.sequence) {
      acc = runTrain(step, acc);
    }
    return { selectedTrainId: resolution.trainId, artifacts: acc };
  }
}

function runTrain(step: string, acc: Record<string, unknown>): Record<string, unknown> {
  if (step === "match") return runMatch(new Map()) as unknown as Record<string, unknown>;
  return acc;
}

File: official/atdd.extension.coder.train-interlocking/implementations/interlocking-infrastructure/fixtures/dirty_direct_wagon/python/trains/interlocking.py
"""InterlockingRunner — DIRTY (direct wagon execution) fixture.

Runner + resolve_train + structured InterlockingResolution all present and the Station Master is
wired, but the runner executes wagons itself: it imports a wagon module, calls run_train(...), and
loops over train.sequence. No Cargo bleed. Only coder.train.interlocking-delegates-to-trainrunner
should fire.
"""
from dataclasses import dataclass

from match.wagon import run_match  # direct wagon import (forbidden)


@dataclass(frozen=True)
class InterlockingResolution:
    interlocking_id: str
    route_id: str
    train_id: str
    train_path: str
    category: str
    category_digit: str
    guard_id: str
    reason: str


class InterlockingRunner:
    def __init__(self, interlocking_yaml_path):
        self._path = interlocking_yaml_path

    def resolve_train(self, action, inputs, state=None):
        return InterlockingResolution(
            interlocking_id="interlocking:match-resolution",
            route_id="nominal-all-voted",
            train_id="3001-solo-match-complete",
            train_path="plan/_trains/3001-solo-match-complete.yaml",
            category="nominal",
            category_digit="3",
            guard_id="all-voted",
            reason="all participants voted",
        )

    def execute(self, action, inputs, state=None, train_runner=None):
        train = _load_train(self.resolve_train(action, inputs, state).train_path)
        result = inputs
        for step in train.sequence:  # loops over train.sequence as executor (forbidden)
            result = run_train(step, result)  # calls run_train directly (forbidden)
        return run_match(result)


def run_train(step, data):
    return data


def _load_train(path):
    class _T:
        sequence = ["match"]

    return _T()
